### ISSUE TITLE
Mutation input object resolving not working

### ISSUE THREAD

**Issue by anluoma** [microsoft/cppgraphqlgen#80]
Hi,

I am trying to create a mutation graphql api with similar input object as below.

`type Mutation {
  addSomething(input: SomeInput!): SomeReturnValue!
}`

Schemagen works ok and objects get created and input struct is also created.

But when running/using the mutation from altair client I can't get it to work. I have used similar query as below (with variables defined).

`mutation inputs($data: SomeInput!){
  addSomething(input: $data)
      {
        field1
     }
}`

Errow what I am getting is:

` "message": "Field error name: addSomething unknown error: Invalid call to Value::find for MapType"
`
which is coming form the first require() call on created convert() method.

Should this work? There is similar thing done in samples for _CompleteTaskInput_

**Comment by wravery** [comment]
Yes, I think that should work. I think the problem may be that the `data` variable is not a properly formed "Map" value. Can you share how you're building that value and setting it on the variables parameter?

**Comment by anluoma** [comment]
thanks for answering. Hmm. I haven't done any special stuff and the whole query is something like this:
```

mutation inputs($data: SomeInput!){
 addSomething(input: $data)
      {
        someId
      }
}


Variables like this:

{
  "data": {
      "someId": "01",
      "somefield": "foobar",
	  }
	  
}
```

I did try without variables also but then it was always complaining that I should add '}' to query. I Tried many ways.. 

Also the exception on cpp code is something I don't understand enough (as don't understand the library's internal logic that much). The code is generated with schemagen and it's  not working on the next lines:
```
std::future<response::Value> Mutation::resolveSomething(service::ResolverParams&& params)
{
	// FIRST THIS IS CALLED.
	auto &argInput = service::ModifiedArgument<SomeInput>::require("input", params.arguments);
	auto result = applySomething(service::FieldParams(params, std::move(params.fieldDirectives)), std::move(argInput));

	return service::ModifiedResult<Something>::convert(std::move(result), std::move(params));
}

This is generated method by schemangen , which gets called after require() calls convert() second time
SomeInput ModifiedArgument<SomeInput>::convert(const response::Value& value)
{
	const auto defaultValue = []()
	{
		response::Value values(response::Type::Map);
		response::Value entry;

		entry = response::Value(true);
		values.emplace_back("isApproved", std::move(entry));

		return values;
	}();

    // THEN Exception in here (inside) that it not a map value.
	auto valueSomeId = service::ModifiedArgument<response::StringType>::require("someId", value)
	
```
I have few queries done on top of this and they seem to work quite nice. Also a question: Do I need to map those variables somehow to the params object by myself or from params object? They are passed to resolve function initially as  response::Value.

**Comment by wravery** [comment]
I just built a quick sample app based on the schema/query you described, check out [wravery/mutation-test](/wravery/mutation-test). I wasn't able to reproduce the error you're seeing though, in fact inline input objects worked fine for me (see the `queryWithInlineInput` version).

This is the output I get from this program:

```json
{"data":{"withVariables":{"somefield":"foobar"}}}
{"data":{"withInlineInput":{"somefield":"foobar"}}}
```

Re: the last question, the variables should be passed as a single `response::Value(response::Type::Map)` object with each of the variable names (minus the `$`) as a key pointing to the `response::Value` for that variable. That object should be passed (as an r-value reference) to the `Request::resolve`/`Operations::resolve` method, and it'll automatically feed them in to the resolvers.

**Comment by anluoma** [comment]
Thanks again for your answer and code. I found some bug in my code related to mapping of those variables. I had done simple string/integer mapping but since this was object that didn't work in this case.

And now seeing your code there seems to be a ready made function response::parseJSON which seems to work better, so thank you!!

One issue (other, should I add another bug?)) still though which I have observed before as well and now tried it with your example. If I add scalar type DateTime to the schema and run example it will crash.
I have had similar things when using scalar type. Strings work fine. Should the schema add some automatic conversion callback for those or how this is supposed to be done?

following schema I tried:
```
schema {
    mutation: Mutation
}
scalar DateTime

input SomeInput {
    isApproved: Boolean! = true
    someId: String!
    somefield: String!
    date: DateTime!
}

type Something {
    someId: String!
    somefield: String!
}

type Mutation {
    addSomething(input: SomeInput): Something!
}


```

**Comment by wravery** [comment]
I'm not sure where exactly that crash is coming from, it looks like maybe a double-free of the `response::Value` object. I wasn't able to reproduce that crash myself, but see the scalar example I just added to [wravery/mutation-test](/wravery/mutation-test).

Generally the service implementer is responsible for serializing the scalar types, they're opaque and can technically can be any valid GraphQL input value or response on output (i.e. JSON). They don't even have to match with the same format in/out AFAICT.

I originally limited input scalar types to objects (`response::Type::Map`) because the examples I saw were all complex types like geo-coordinates or date/time. If you pass `null` or a plain string it'll throw an exception in [`ModifiedArgument<response::Value>::convert`](https://github.com/microsoft/cppgraphqlgen/blob/master/src/GraphQLService.cpp#L595), but now that I think about it, there's no reason it couldn't be any valid GraphQL input type. It should be up to the service implementation how to interpret the `response::Value` that it receives as an argument.
